tiledb/sm/query/readers/dense_reader.cc

```diff
@@ -783,94 +824,131 @@ Status DenseReader::apply_query_condition(
     }
 
     // Read and unfilter query condition attributes.
-    RETURN_CANCEL_OR_ERROR(
-        read_and_unfilter_attribute_tiles(qc_names, result_tiles));
+    std::vector<FilteredData> filtered_data =
+        read_attribute_tiles(qc_names, result_tiles);
 
-    if (stride == UINT64_MAX) {
-      stride = 1;
+    if (compute_task.valid()) {
+      RETURN_NOT_OK(storage_manager_->compute_tp()->wait(compute_task));
     }
 
-    // Process all tiles in parallel.
-    auto status = parallel_for_2d(
-        storage_manager_->compute_tp(),
-        t_start,
-        t_end,
-        0,
-        num_range_threads,
-        [&](uint64_t t, uint64_t range_thread_idx) {
-          // Find out result space tile and tile subarray.
-          const DimType* tc = (DimType*)&tile_coords[t][0];
-          auto it = result_space_tiles.find(tc);
-          assert(it != result_space_tiles.end());
-
-          // Iterate over all coordinates, retrieved in cell slab.
-          const auto& frag_domains = it->second.frag_domains();
-          TileCellSlabIter<DimType> iter(
-              range_thread_idx,
+    compute_task = storage_manager_->compute_tp()->execute(
+        [&,
+         filtered_data = std::move(filtered_data),
+         qc_names,
+         t_start,
+         t_end,
+         num_range_threads,
+         result_tiles]() {
+          // For easy reference.
+          const auto& tile_coords = subarray.tile_coords();
+          const auto dim_num = array_schema_.dim_num();
+          auto stride = array_schema_.domain().stride<DimType>(layout_);
+          const auto cell_order = array_schema_.cell_order();
+          const auto global_order = layout_ == Layout::GLOBAL_ORDER;
+
+          // Unfilter tiles.
+          for (auto& name : qc_names) {
+            RETURN_NOT_OK(unfilter_tiles(name, result_tiles));
+          }
+
+          if (stride == UINT64_MAX) {
+            stride = 1;
+          }
+
+          // Process all tiles in parallel.
+          auto status = parallel_for_2d(
```

Comment:
Do we need to use `num_range_threads-1` here to count for the extra compute thread?

Reply:
The range threads are not actually the number of threads in the threadpool. They are only used when a read will be a few large tiles. At that point, we will split the work for a tile across threads. Also, by the time the work of the parallel for in the compute task gets processed, the compute task should already mostly be in a waiting/yielding state.